lib/proxy/peer/credentials.go

```diff
@@ -0,0 +1,207 @@
+// Copyright 2022 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package peer
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"net"
+
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc/credentials"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/teleport/lib/tlsca"
+)
+
+// serverCredentials wraps a [crendentials.TransportCredentials] that
+// extends the ServerHandshake to ensure the credentials contain the proxy system role.
+type serverCredentials struct {
+	credentials.TransportCredentials
+}
+
+// newServerCredentials creates new serverCredentials from the given [crendentials.TransportCredentials].
+func newServerCredentials(creds credentials.TransportCredentials) *serverCredentials {
+	return &serverCredentials{
+		TransportCredentials: creds,
+	}
+}
+
+// ServerHandshake performs the TLS handshake and then verifies that the client
+// attempting to connect is a Proxy.
+func (c *serverCredentials) ServerHandshake(conn net.Conn) (_ net.Conn, _ credentials.AuthInfo, err error) {
+	conn, authInfo, err := c.TransportCredentials.ServerHandshake(conn)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	defer func() {
+		if err != nil {
```

Comment:
Are you sure this will do you what you want? 

The error from `checkProxyRole` below, for example, will not overwrite the `err` variable you're checking here.

Reply:
anywhere you use `if err :=` introduces a new scope with a different `err` than the named return, but returning `trace.Wrap(err)` implicitly assigns to the named return so it should work.

Reply:
Why are we using named returns to begin with? The only one we name is `err` and that's kind of obvious.

Reply:
I see. Can you just confirm that Gavin's correct that the implicit assignment to the named return happens before the deferred func evaluates?

Reply:
@zmb3 I tested that it does with go playground before I commented, just to be sure since named returns can be subtle: https://go.dev/play/p/dUn1k3lnifJ